File: output/deepseek_r1/avatar/algo_based_translation/python/java/successful_translation/atcoder_ABC145_E.java

import java.util.*;
import java.io.*;

public class atcoder_ABC145_E {
    static class Pair {
        int time;
        int value;
        Pair(int time, int value) {
            this.time = time;
            this.value = value;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int T = Integer.parseInt(st.nextToken());

        Pair[] timeValue = new Pair[N + 1];
        timeValue[0] = new Pair(-1, -1);
        for (int i = 1; i <= N; i++) {
            st = new StringTokenizer(br.readLine());
            timeValue[i] = new Pair(Integer.parseInt(st.nextToken()), Integer.parseInt(st.nextToken()));
        }

        Arrays.sort(timeValue, 1, N + 1, Comparator.comparingInt(p -> p.value));

        int[][] dp = new int[N + 1][T];
        for (int[] row : dp) Arrays.fill(row, -1);
        for (int t = 0; t < T; t++) dp[0][t] = 0;
        for (int n = 1; n <= N; n++) dp[n][0] = 0;

        for (int n = 1; n <= N; n++) {
            for (int t = 1; t < T; t++) {
                if (timeValue[n].time > t) {
                    dp[n][t] = dp[n - 1][t];
                } else {
                    int prev = t - timeValue[n].time;
                    dp[n][t] = Math.max(dp[n - 1][t], timeValue[n].value + dp[n - 1][prev]);
                }
            }
        }

        int valAcum = timeValue[N].value;
        int tRemaining = T - 1;
        int maxVal = valAcum + dp[N - 1][tRemaining];

        for (int n = N - 1; n >= 1; n--) {
            valAcum += timeValue[n].value;
            tRemaining -= timeValue[n + 1].time;
            if (tRemaining < 0) break;
            maxVal = Math.max(maxVal, valAcum + dp[n - 1][tRemaining]);
        }

        System.out.println(maxVal);
    }
}

File: output/deepseek_r1/avatar/algo_based_translation/java/python/successful_translation/atcoder_ABC145_E.py

import sys

def main():
    N, T = map(int, sys.stdin.readline().split())
    AB = [tuple(map(int, sys.stdin.readline().split())) for _ in range(N)]
    AB.sort(key=lambda x: x[0])
    dp = [[0] * 6001 for _ in range(N+1)]
    
    for i in range(1, N+1):
        time, value = AB[i-1]
        for time_after in range(6001):
            dp[i][time_after] = dp[i-1][time_after]
            if time_after - time >= 0 and time_after - time < T:
                dp[i][time_after] = max(dp[i][time_after], dp[i-1][time_after - time] + value)
    
    print(max(dp[N]))

if __name__ == "__main__":
    main()

